### ISSUE TITLE
Memory Transfer Overhead in ppc64Ie

### ISSUE THREAD

**Issue by ppnaik1890** [torch-spyre/torch-spyre#3255]
## Summary

Across the benchmark report, **14 out of 16 ops (as per #1643) show higher memory transfer time on ppc64le** compared to x86_64. 

### Critical (3.5x–4x slower on ppc64le)
- Granite full-model configs (both prefill and decode) show the worst regression
- Granite one-block prefill config is similarly affected

### High (2x–3x slower on ppc64le)
- Granite decode one-block config
- All matmul ops with large inner dimension (K=12800)
- Both mlp shapes (prefill and decode)
- Small-shape pointwise_add and softmax (batch=4, seq_len=1)
- Matmul with medium shapes (K=4096, batch=4)

### Moderate (1x–1.7x slower on ppc64le)
- Matmul with smaller output dimension (K=1024)
- Matmul prefill shape with K=4096

### Not Affected / Faster on ppc64le
- Attention op (prefill shape) — actually faster on ppc64le
- Large-shape pointwise_add, softmax, layernorm (seq_len=512)

## Patterns Observed

1. **Decode shapes are worse than prefill shapes** — small batch/seq_len ops consistently show higher regression ratios than their larger counterparts within the same op type.
2. **Regression scales with tensor size** — matmul ops with larger weight matrices (12800 > 4096 > 1024) show progressively worse ratios.
3. **Large prefill shapes are less affected** — ops with seq_len=512 tensors show minimal or no regression, suggesting the issue is more pronounced when transfer overhead dominates over compute.

## Hypotheses to Investigate

1. **DMA/PCIe configuration differences** — ppc64le may have suboptimal DMA transfer settings or different PCIe topology introducing extra latency per transfer.
2. **Small transfer penalty** — the pattern of decode shapes being worse suggests a per-transfer fixed overhead that is higher on ppc64le, disproportionately affecting small payloads.
3. **Memory alignment or page size mismatch** — ppc64le uses 64KB pages by default vs 4KB on x86_64; if transfer buffers are not aligned to device expectations, this could cause fragmented transfers.

## Next Steps

- [ ] Reproduce with minimum repro code to confirm/highlight the issue.
- [ ] Analyze the difference further using the fine grain perf report and profiler trace files (will provide memset vs D2H H2D transfer overheads)

**Comment by bjking1** [comment]
Is this memory transfer between the host and the Spyre card? In other words, PCIe? If so, what speed is x86 running at? On Power, we are running in a Gen4 I/O drawer, so are limited to Gen4x16 per card. In fact, we are limited to Gen4x16 for a group of 4 cards under a switch since we have a single Gen4x16 cable card going to the group of 4.

**Comment by kiruthikavaradarajan** [comment]
Raghavendra & Avinash is looking into it.

**Comment by raghu-gj** [comment]
TRACE-level logging was enabled and a bpftrace script was used to probe the flex and senlib libraries during a matmul benchmark run:

`python run_benchmark.py --op matmul --shape 1 512 4096 --shape 4096 12800 --stacks torch-spyre --run 2`
The tensor under test has shape [4096 × 12800] — 52.4 M elements, approximately 100 MiB in FP16/BF16.

**Findings**
**1. CB Interface (Control Block) — one-time cost, not a concern**

At startup, `RegisterMemForIOMMU(cq, cq_size)` and `pin_and_map `are called once and the result is stored in `cq_pm_.` On each subsequent matmul, the CB path performs only a plain memcpy into `queue_ptr_[idx]` and a single MMIO write via `TapQueue()`. There is zero IOMMU overhead per operation on this path.

**2. Weight tensor transfer — the source of per-op latency**

The 100 MiB weight matrix lives in PyTorch host RAM and must be transferred to the AIU over PCIe via the VFIO backend. The transfer is broken into fixed 4 MiB chunks controlled by two constants in the flex code:

`FLEX_DATA_TRANSFER_BUFFER_SIZE = 4 MiB — size of each DMA chunk`
`FLEX_DATA_TRANSFER_BUFFER_COUNT = 8 — max in-flight chunks before the pipeline stalls`
For a 100 MiB tensor, this produces 25 chunks. Every chunk independently calls` vfio_backend::RegisterMemForIOMMU() `and `pin_and_map(hmva, size_bytes)`, followed by an `unmap` after the transfer completes. This means 25 DMA map + 25 DMA unmap calls per matmul operation.

Note: on x86, `RegisterMemForIOMMU` itself is a no-op `((void)addr).` The measured latency comes entirely from the `pin_and_map `/ VFIO DMA mapping path.

**3. Pipeline behaviour**

The chunks are pipelined, not parallel. Chunks 0–7 are submitted immediately, each getting a fresh DmaBuf object. From chunk 8 onward, the submission loop blocks until one of the in-flight buffers completes and is recycled. This sliding-window behaviour means map/unmap cycles repeat for every chunk beyond the first 8, keeping the per-op DMA mapping cost proportional to tensor size.

**Root Cause**
Each 4 MiB chunk of payload data is dynamically DMA-mapped and unmapped on every operation. For a 100 MiB weight tensor this results in **25 map + 25 unmap** calls per matmul, which is the primary source of the additional latency observed.

**Next Steps**
1. Quantify map/unmap call counts per operation. Use the  bpftrace tooling to capture register and unregister map counts for each op to establish a precise latency baseline per DMA cycle. 
2.  Try to reduce map/unmap calls by reusing existing mappings
